cpp/autosar/src/rules/A16-0-1/PreProcessorShallOnlyBeUsedForCertainDirectivesPatterns.ql

```diff
@@ -21,13 +21,26 @@ import cpp
 import codingstandards.cpp.autosar
 import codingstandards.cpp.FunctionLikeMacro
 
+class PermittedInnerDirectiveType extends PreprocessorDirective {
+  PermittedInnerDirectiveType() {
+    //permissive listing for directives that can be used in a valid wrapper
+    this instanceof MacroWrapper or
+    this instanceof PreprocessorEndif or
+    this instanceof Include or
+    this instanceof PermittedMacro or
+    this instanceof PreprocessorElif or
```

Comment:
To consider `elseif` separately we need to change the `getAGuard` predicate to slice up the range into appropriate ranges. That is from `if` to `elseif` and from `elseif` to `elseif` or `endif`.

We currently aren't capturing the case:

```
#ifdef MACRO_ENABLED_NON  // COMPLIANT
#include <string>         // COMPLIANT
#elif MACRO_ENABLED_OTHER // COMPLIANT
int x = 1;                // NON_COMPLIANT
#endif     
```

Reply:
this one is complicated by the fact that the database will or will not have an element in it based on the outcome of the preprocessor conditional

I built a full enumeration of the testcases that cover all scenarios

however if I add this to the testcase, it has defs and undefs that make it into the alerts
unsure if it should still be added for explanation sake though?

```

// case 1 - first present only
#define MACRO_ENABLED_NON 1
#ifdef MACRO_ENABLED_NON  // COMPLIANT
#include <string>         //present
#elif MACRO_ENABLED_OTHER // NON_COMPLIANT[FALSE_NEGATIVE]
int x = 1;  // not present
#endif

// case 2 - second present only
#undef MACRO_ENABLED_NON
#define MACRO_ENABLED_OTHER 1
#ifdef MACRO_ENABLED_NON  // COMPLIANT
#include <string>         //not present
#elif MACRO_ENABLED_OTHER // NON_COMPLIANT
int x = 1;  // present
#endif

// case 3 - neither present
#undef MACRO_ENABLED_NON
#undef MACRO_ENABLED_OTHER
#ifdef MACRO_ENABLED_NON  // COMPLIANT
#include <string>         //not present
#elif MACRO_ENABLED_OTHER // NON_COMPLIANT[FALSE_NEGATIVE]
int x = 1;  // not present
#endif

// case 4 - both look present but the second still not bc the condition is not
// required to be evaluated
#define MACRO_ENABLED_NON 1
#define MACRO_ENABLED_OTHER 1
#ifdef MACRO_ENABLED_NON  // COMPLIANT
#include <string>         //present
#elif MACRO_ENABLED_OTHER // NON_COMPLIANT[FALSE_NEGATIVE]
int x = 1;  // not present
#endif
```

I have made an adjustment to the guard range logic though to avoid duplication where possible
so let me know if that now makes more sense

Reply:
additionally now I am a little unsure of how to label this testcase
```
#ifdef OBJECTLIKE_MACRO_NO // COMPLIANT
int x0 = 0;                // not present
#elif OBJECTLIKE_MACRO > 0 // NON_COMPLIANT
int x0 = 1; // present
#endif                     // COMPLIANT
```
according to the examples in the spec `#ifdef OBJECTLIKE_MACRO_NO // COMPLIANT` should maybe be `#ifdef OBJECTLIKE_MACRO_NO // NON_COMPLIANT[FALSE_NEGATIVE] - known due to database not containing elements`

Reply:
I think it would be good to have the test cases in for documentation purposes.
To prevent the `#define` and `#undef` to be part of the results set you can use an option file to set the preprocessor defines like you would on a compiler invocation. You do have to change the cases because you can't define and undefine on the command line.

Yes, I agree that `OBJECTLIKE_MACRO_NO` is `// NON_COMPLIANT[FALSE_NEGATIVE] - known due to database not containing elements`

Reply:
awesome, thanks, didnt think of/know(?) about the command line flags, excellent suggestion!